File: node.hs
data Node = Node {
  value :: Int,
  left :: Maybe Node,
  right :: Maybe Node
}

sumTree :: Node -> Int
sumTree (Node value Nothing Nothing) = value
sumTree (Node value (Just left) Nothing) = value + sumTree left
sumTree (Node value Nothing (Just right)) = value + sumTree right
sumTree (Node value (Just left) (Just right)) = value + sumTree left + sumTree right

leftNode :: Node -> Maybe Node
leftNode (Node _ l _) = l
main :: IO ()
main = do
  let tree = Node 1 (Just (Node 2 Nothing Nothing)) (Just (Node 3 Nothing Nothing))
  let total = sumTree tree -- 6
  let tree2 = tree { right = Nothing } -- tree2 is now Node 1 (Just (Node 2 Nothing Nothing)) Nothing
  let tree3 = leftNode tree -- tree3 is now Just (Node 2 Nothing Nothing)
  print total

File: node.rs
struct Node {
    value: i32,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

fn sum(node: &Node) -> i32 {
    let mut acc = node.value;
    if let Some(left) = &node.left {
        acc += sum(left);
    }
    if let Some(right) = &node.right {
        acc += sum(right);
    }
    acc
}

fn main() {
    let mut tree = Node {
        value: 1,
        left: Some(Box::new(Node {
            value: 2,
            left: None,
            right: None,
        })),
        right: Some(Box::new(Node {
            value: 3,
            left: None,
            right: None,
        })),
    };
    let total = sum(&tree); // 6
    tree.right = None; // tree is now Node { value: 1, left: Some(...), right: None }
    let left = tree.left;
    let unwrapped = left.unwrap();
    tree = *unwrapped; // tree is now Node { value: 2, left: None, right: None }
}
